design/cassandra/design.md

```diff
@@ -0,0 +1,186 @@
+
+|  Cassandra    |  Kubernetes   |
+|   ----        |     ----      |
+|    Cluster    |  Cluster CRD  |
+|   Datacenter  |  StatefulSet  |
+|     Node      |     Pod       |
+
+### Cassandra Cluster CRD
+
+An example CRD would look something like this: 
+
+``` yaml
+apiVersion: "cassandra.rook.io/v1alpha1"
+kind: "Cluster"
+metadata:
+  name: "my-cassandra-cluster"
+spec:
+  version: "3.1.11"
+  # Optional: repository overrides the default image repo
+  repository: "custom-enterprise-repo.io/cassandra"
+  dataCenter:
+    name: "us-east-1"
+    racks:
+    - name: "us-east-1c"
+      instances: 3
+      # Optional: configMapName references a user's custom configuration for
+      # a specific Cassandra Rack 
+      configMapName: "cassandra-config"
+      # Rook Common Type: StorageSpec
+      storage:
+        volumeClaims:
+          - storageClassName:
+            metadata:
+              name: "cassandra-data"
+            spec:
+              accessModes: ["ReadWriteOnce"]
+              storageClassName: default
+              resources:
+                requests:
+                  storage: "500Gi"
+      # Rook Common Type: PlacementSpec
+      # Optional: Placement declares node/pod (anti)affinity
+      placement:
+        nodeAffinity: 
+          requiredDuringSchedulingIgnoredDuringExecution:
+            nodeSelectorTerms:
+              - matchExpressions:
+                - key: failure-domain.beta.kubernetes.io/region
+                  operator: In
+                  values:
+                  - us-east-1
+                - key: failure-domain.beta.kubernetes.io/zone
+                  operator: In
+                  values:
+                  - us-east-1c
+      # Rook Common Type: ResourceSpec
+      # Resources declares the CPU,RAM resources of a single instance of the Rack.
+      resources:
+        - name: cassandra 
+          requests:
+            cpu: "2000m"
+            memory: "4Gi"
+          limits:
+            cpu: "2000m"
+            memory: "4Gi"
+        - name: sidecar
+          requests:
+            cpu: "200m"
+            memory: "200M"
+          limits:
+            cpu: "200m"
+            memory: "200M"
+  # Optional: sidecarImage overwrites the default image used
+  sidecarImage: 
+    repository: "rook.io/sidecar-cassandra"
+    tag: "0.1"
+```
+
+* The operator will create a StatefulSet for each (Datacenter, Rack) tupple.
+* In practice, Datacenter is usually mapped to a Region and Rack to an Availability Zone.
+
+### Major Pain Point: Stable Pod Identity
+
+A problem that every approach so far has, is dealing with loss of persistence.
+If a Cassandra instance loses its data (ie the underlying node fails, the disk dies).
+The replacing instance has to provide the IP of the instance it's replacing. Cassandra doesn't support stable hostnames as an identification method.
+However, IPs on Kubernetes are ephemeral and that leaves us with 2 options:
+
```

Comment:
Part of this problem can be solved by using persistent volume claims and mapping to stable names/labels provided by the stateful sets. Obviously, this just shifts the blame around, however, in the general case this allows to proverbially save the data which leaves us with only having to map IP addresses. Ie replace them which brings us to your bookkeeping point.
In any case, having the option of PVC vs Local Discs is a sensible solution that can be left as a setting to the end user IMHO but for now, we can avoid feature creep till we have something stable.